modules/core/src/main/java/org/apache/ignite/internal/processors/affinity/GridAffinityAssignmentV2.java

```diff
@@ -0,0 +1,334 @@
+    /** */
+    private static final long serialVersionUID = 0L;
+
+    /** Topology version. */
+    private AffinityTopologyVersion topVer;
+
+    /** Collection of calculated affinity nodes. */
+    private List<List<ClusterNode>> assignment;
+
+    /** Map of primary node partitions. */
+    private Map<UUID, Set<Integer>> primary;
+
+    /** Map of backup node partitions. */
+    private Map<UUID, Set<Integer>> backup;
+
+    /** Assignment node IDs */
+    private transient volatile List<Collection<UUID>> assignmentIds;
+
+    /** Nodes having primary or backup partition assignments. */
+    private transient volatile Set<ClusterNode> nodes;
+
+    /** Nodes having primary partitions assignments. */
+    private transient volatile Set<ClusterNode> primaryPartsNodes;
+
+    /** */
+    private transient List<List<ClusterNode>> idealAssignment;
+
+    /**
+     * Default constructor for deserialization.
+     */
+    public GridAffinityAssignmentV2() {
+
+    }
+
+    /**
+     * Constructs cached affinity calculations item.
+     *
+     * @param topVer Topology version.
+     */
+    GridAffinityAssignmentV2(AffinityTopologyVersion topVer) {
+        this.topVer = topVer;
+        primary = Collections.emptyMap();
+        backup = Collections.emptyMap();
+    }
+
+    /**
+     * @param topVer Topology version.
+     * @param assignment Assignment.
+     * @param idealAssignment Ideal assignment.
+     */
+    public GridAffinityAssignmentV2(AffinityTopologyVersion topVer,
+        List<List<ClusterNode>> assignment,
+        List<List<ClusterNode>> idealAssignment
+    ) {
+        assert topVer != null;
+        assert assignment != null;
+        assert idealAssignment != null;
+
+        this.topVer = topVer;
+        this.assignment = Collections.unmodifiableList(assignment);
+        this.idealAssignment = Collections.unmodifiableList(
+            idealAssignment.equals(assignment) ? assignment : idealAssignment
+        );
+
+        // Temporary mirrors with modifiable partition's collections.
+        Map<UUID, Set<Integer>> tmpPrimary = new HashMap<>();
+        Map<UUID, Set<Integer>> tmpBackup = new HashMap<>();
+        boolean isPrimary;
+
+        for (int partsCnt = assignment.size(), p = 0; p < partsCnt; p++) {
+            isPrimary = true;
+
+            for (ClusterNode node : assignment.get(p)) {
+                UUID id = node.id();
+
+                Map<UUID, Set<Integer>> tmp = isPrimary ? tmpPrimary : tmpBackup;
+
+                /*
+                    https://issues.apache.org/jira/browse/IGNITE-4554 BitSet performs better than HashSet at most cases.
+                    However with 65k partition and high number of nodes (700+) BitSet is loosing HashSet.
+                    We need to replace it with sparse bitsets.
+                 */
+                tmp.computeIfAbsent(id, uuid ->
+                    !IGNITE_DISABLE_AFFINITY_MEMORY_OPTIMIZATION ? new BitSetIntSet() : new HashSet<>()
```

Comment:
Missing 
partAssignments = assignment.get(p)
U.newHashSet(partAssignments.size())

Reply:
You can't know for sure the size of hashSet and this is no partAssignments.size().